sdk/src/Core/Amazon.Runtime/Pipeline/Handlers/BaseAuthResolverHandler.cs

```diff
@@ -65,40 +66,64 @@ protected void PreInvoke(IExecutionContext executionContext)
-                    executionContext.RequestContext.Identity = clientConfig.DefaultAWSCredentials;
-                    break;
-                }
+                    executionContext.RequestContext.Signer = GetSigner(scheme);
 
-                if (scheme is BearerAuthScheme && clientConfig.AWSTokenProvider != null)
-                {
-                    // If the legacy token provider is set, we'll use it to resolve the identity.
-#if NETFRAMEWORK
-                    var resolvedToken = clientConfig.AWSTokenProvider.TryResolveToken(out var token);
-                    if (!resolvedToken)
+                    if ((scheme is AwsV4aAuthScheme || scheme is AwsV4AuthScheme) && clientConfig.DefaultAWSCredentials != null)
                     {
-                        continue;
+                        // We can use DefaultAWSCredentials if it was set by the user for these schemes.
+                        executionContext.RequestContext.Identity = clientConfig.DefaultAWSCredentials;
+                        return;
                     }
+
+                    if (scheme is BearerAuthScheme && clientConfig.AWSTokenProvider != null)
+                    {
+                        // If the legacy token provider is set, we'll use it to resolve the identity.
+#if NETFRAMEWORK
+                        var resolvedToken = clientConfig.AWSTokenProvider.TryResolveToken(out var token);
+                        if (!resolvedToken)
+                        {
+                            continue;
+                        }
 #else
-                    var resolvedToken = clientConfig.AWSTokenProvider.TryResolveTokenAsync().GetAwaiter().GetResult();
-                    if (!resolvedToken.Success)
+                        var resolvedToken = clientConfig.AWSTokenProvider.TryResolveTokenAsync().GetAwaiter().GetResult();
+                        if (!resolvedToken.Success)
+                        {
+                            continue;
+                        }
+
+                        var token = resolvedToken.Value;
+#endif
+
+                        executionContext.RequestContext.Identity = token;
+                        return;
+                    }
+
+                    var identityResolver = scheme.GetIdentityResolver(clientConfig.IdentityResolverConfiguration);
+                    executionContext.RequestContext.Identity = identityResolver.ResolveIdentity();
+
+                    if (executionContext.RequestContext.Identity != null)
+                    {
+                        return;
+                    }
+                }
+                catch (Exception)
+                {
+                    // If there are multiple authentication schemes and we cannot resolve the identity for some reason (e.g. the CRT is
+                    // required for SigV4A signing), we'll attempt the next option before returning.
+                    if (authOptions.Count > 1)
```

Comment:
From a theoretical point of view we should be checking if there are more authOptions to evaluate not that there are just more then one. In theory what if there were 2 authOptions and the first one cleanly didn't resolve a token so a `continue` happened and then the second one threw an exception. We would exit this loop with no exception and no identity.

Reply:
Can we add an SDK debug logging here?

Reply:
Addressed both comments in https://github.com/aws/aws-sdk-net/pull/3622/commits/4e6589d34a4ac75a084ea9a17c412562f084afdb